### PR TITLE
Add types for `chai-json-pattern`

### PR DESCRIPTION
**Pull request by chrishrb**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/chai-json-pattern/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        [key: string]: (target: any) => void;
+    }
+
+    interface ChaiJsonPattern extends Chai.ChaiPlugin {
+        extend(plugin: Plugin): void;
+    }
+}
+
+declare const chaiJsonPattern: ChaiJsonPattern.ChaiJsonPattern;
+export = chaiJsonPattern;
```
Comment (sheetalkamat) [inline types/chai-json-pattern/index.d.ts:28]: Use export default to export things from module since that's what package does

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: types/chai-json-pattern/index.d.ts (hunk 1 above)
Comment (chrishrb) [reply types/chai-json-pattern/index.d.ts]: should be resolved now, thanks 👍

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## chai-json-pattern ([<kbd>unpkg</kbd>](https://unpkg.com/browse/chai-json-pattern@latest/))
was missing the following properties:
1. extend
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 60c669711f3d859df7e545f561483b00680aa27d
</p>

**Comment by chrishrb** [comment]
> Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files. The check for missing properties isn't always right, so take this list as advice, not a requirement.
> 
> ## chai-json-pattern ([unpkg](https://unpkg.com/browse/chai-json-pattern@latest/))
> was missing the following properties:
> 
> 1. extend
> 
> Generated by 🚫 [dangerJS](https://danger.systems/js) against [60c6697](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/60c669711f3d859df7e545f561483b00680aa27d)

hmm, in the tests the `extend` function was resolved correctly - maybe it's indeed wrong here?

File: types/chai-json-pattern/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        [key: string]: (target: any) => void;
+    }
+
+    interface ChaiJsonPattern extends Chai.ChaiPlugin {
+        extend(plugin: Plugin): void;
+    }
+}
+
+declare const chaiJsonPattern: ChaiJsonPattern.ChaiJsonPattern;
+export default chaiJsonPattern;
```
Comment (sheetalkamat) [inline types/chai-json-pattern/index.d.ts:28]:
```suggestion
}
declare const chaiJsonPattern: Chai.ChaiPlugin;
export default chaiJsonPattern;
export function extend(plugin: ChaiJsonPattern.Plugin): void;
```

Comment (sheetalkamat) [reply types/chai-json-pattern/index.d.ts:28]: There is no `extend` method on default object
````
const a = require("chai-json-pattern");
console.log(a.default);
console.log(a.default.extend);
console.log(a.extend);
c:\temp\test>node a.js
[Function: chaiJsonPattern]
undefined
[Function: extend]
```

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: types/chai-json-pattern/index.d.ts (hunk 2 above)
Comment (chrishrb) [reply types/chai-json-pattern/index.d.ts]: hm, i don‘t think this is want we want here. I don‘t want to change the api, only add the correct types but if we change the export here like you suggested then I have to change the import like this?

```ts
import chaiJsonPattern, { extend } from 'chai-json-pattern';
```

In the javascript source code the API for the extend method is described like [this](https://github.com/damian-szulc/chai-json-pattern/blob/master/API.md#extendextension) and i want to keep this if possible. I think the original PR was the right choice for keeping this API or am I completely wrong here?

Comment (chrishrb) [reply types/chai-json-pattern/index.d.ts:28]: nvm. you are right, i will fix this

**Review by RyanCavanaugh — APPROVED** [review]

**Pull request merged**